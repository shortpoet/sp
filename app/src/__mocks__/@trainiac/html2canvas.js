
// const html2canvas = jest.genMockFromModule('@trainiac/html2canvas')

console.log('hello from html2canvas mock')

const html2canvas = jest.fn(() => {
  return {
    width: 888,
    height: 888
  }
})

module.exports = html2canvas