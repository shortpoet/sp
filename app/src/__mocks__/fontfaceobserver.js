
const fontfaceobserver = jest.genMockFromModule('fontfaceobserver')

console.log('hello from fontfaceobserver mock')

module.exports = fontfaceobserver