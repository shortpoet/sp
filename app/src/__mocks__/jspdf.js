// these modules' names must match exactly.  i had jsPDF at first and it didn't work
// git renamed this file on a merge BUG OPPORTUNITY

/* gotta watch out for how return values are nested
previously 
return {
  doc: {
    internal: ...
  }
}
had one too many {} and was not matching expected in function
*/

console.log('hello from jsPDF mock')

// https://stackoverflow.com/questions/44686077/how-to-use-jest-to-test-file-download
import FileSaver from 'file-saver'
export const fs = jest.mock('file-saver', ()=>({saveAs: jest.fn()}))

const getWidth = jest.fn(() => 888)
const getHeight = jest.fn(() => 888)
export const addImage = jest.fn(() => 'Test Image')
export const addPage = jest.fn(() => 'Test Page')
export const save = jest.fn(() => fs)
const doc = {
  internal: {
    pageSize: {
      getWidth: getWidth,
      getHeight: getHeight
    }
  },
  addImage: addImage,
  addPage: addPage,
  save: save
  // addImage: jest.fn(() => Promise.resolve(true)),
  // save: jest.fn(() => Promise.resolve(true))

}



// class jspdf {
//   constructor (type, size, dimension) {
//     this.internal = {
//       pageSize: {
//         getWidth: getWidth,
//         getHeight: getHeight
//       }
//     }

//   }

//   addImage = () => jest.fn(() => 'Test Image')

//   save = () => jest.fn(() => 'Test Save')

// }

const mock = jest.fn().mockImplementation(() => {
  return doc;
});

export default mock;

// module.exports = jspdf
