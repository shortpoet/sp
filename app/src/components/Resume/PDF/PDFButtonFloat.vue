<template>
  <portal :to="target">
    <!-- 
      * previously had this selector to unify the components - split after styling diverged
      * v-if="href" 
      * v-if="pdfTarget"
    </div>-->
    <div type="input" :class="classObject" @click="showModal">
      <font-awesome-layers class="button-float-icon-layer fa-lg">
        <font-awesome-icon class="button-float-icon-circle" size="2x" icon="circle" />
        <font-awesome-icon
          class="button-float-icon"
          size="2x"
          :transform="_icon.transform"
          :icon="_icon.icon"
        ></font-awesome-icon>
      </font-awesome-layers>
    </div>
    <div class="modal-slot">
      <PDFModal
        v-show="isModalVisible"
        @close="closeModal"
        @to-pdf="savePDF"
        @to-page="toPage"
        @to-canvas="toCanvas"
      >
        <!-- <template v-slot:header>
          <h1>Test Header</h1>
        </template>
        <template v-slot:body>
          <h1>Test Body</h1>
        </template>
        <template v-slot:footer>
          <h1>Test Footer</h1>
        </template>-->
      </PDFModal>
    </div>
  </portal>
</template>

<script>
import PDFModal from "@/components/Resume/PDF/PDFModal";
import { colorLog } from "@/utils/colorLog";
import { log } from "@/utils/colorLog";
// log('blue', 'test')
import jsPDF from "jspdf";
// using fork for now to solve this issue
// the changes i made with the icons as the deployed version still works
// https://github.com/niklasvh/html2canvas/issues/1868#issuecomment-599217709
import html2canvas from "@trainiac/html2canvas";
// import html2canvas from 'html2canvas'
const FontFaceObserver = require("fontfaceobserver");
import moment from "moment";

// https://stackoverflow.com/questions/38975138/is-using-async-in-settimeout-valid
// let wait = ms => new Promise(resolve => setTimeout(resolve, ms));

export default {
  name: "PDFButtonFloat",
  components: {
    PDFModal
  },
  props: {
    target: {
      type: String
    },
    icon: {
      type: String
    },
    pdfTarget: {
      type: String
    }
    // handler: {
    //   type: Function,
    //   required: false
    // }
  },
  data() {
    return {
      iconMap: {
        pdf: {
          icon: "file-pdf",
          transform: "shrink-5 right-2.3"
        },
        save: {
          icon: "save",
          transform: "shrink-5 right-1"
        }
      },
      isModalVisible: false,
      canvas: null,
      jspdf: null
    };
  },
  computed: {
    mobile() {
      return window.innerWidth < 768;
    },
    classObject() {
      return {
        "button-float": true,
        mobile: this.mobile
      };
    },
    _icon() {
      return this.iconMap[`${this.icon}`];
    }
  },
  methods: {
    showModal() {
      this.isModalVisible = true;
    },
    closeModal() {
      this.isModalVisible = false;
    },
    /* istanbul ignore next */
    toCanvas() {
      html2canvas(document.getElementById(this.pdfTarget), {
        allowTaint: true
      }).then(canvas => {
        document.body.appendChild(canvas);
      });
    },
    async checkFonts() {
      var fontA = new FontFaceObserver("Open Sans");
      var fontB = new FontFaceObserver("Saira Extra Condensed");
      // must return otherwise load order is not correct
      try {
        return Promise.all([fontA.load(), fontB.load()]).then(function() {
          console.log("Family A & B have loaded");
        });
      } catch {
      /* istanbul ignore next */
        err => console.log(err);
      }
    },
    async getCanvas(options) {
      const vm = this;
      const target = document.getElementById(vm.pdfTarget);
      try {
        let canvas = html2canvas(target, {
          ...options
        });
        return canvas;
      } catch {
      /* istanbul ignore next */
        err => console.log(err);
      }
    },
    // first one is significantly faster
    // second makes mage unresponsive for like 3 secs
    // go unit tests!

    // TODO maybe add the functionality to provide filetype here
    // and then in addImage
    // then would need to actually test this function
    /* istanbul ignore next */
    getDataURL(canvas) {
      return canvas.toDataURL("image/jpeg", 1.0);
      // return canvas.toDataURL('image/png');
    },
    // https://stackoverflow.com/questions/19626680/is-settimeout-a-good-solution-to-do-async-functions-with-javascript
    async setCanvas(options, callback) {
      const vm = this;
      await vm.checkFonts();
      colorLog("fonts have been checked", "violet");
      setTimeout(async () => {
        try {
          const canvas = await vm.getCanvas(options);
          vm.canvas = canvas;
          callback();
        } catch (err) {
        /* istanbul ignore next */
          console.log(err);
        }
      }, 250);
    },
    // close modal first
    // [Vue warn]: Method "isModalVisible" has type "boolean" in the component definition. Did you reference the function correctly?      this.closeModal()
    // [Vue warn]: Method "isModalVisible" has already been defined as a data property.
    // vm.isModalVisible = false
    // vm.closeModal()
    // both threw those warning but only in test
    // realized that the click event on the modal buttons
    // bubbled up to parent backdrop div that closes on click
    // which is why the modal tests had that double event that had me wondering
    // seems then no need to call close from here which makes sense as far as separation of concerns i guess
    //
    // setting width and height cuts it off at the "one page mark"
    // width: 810,
    // height: 1100,
    toPDF() {
      const vm = this;
      const canvas = vm.canvas;
      this.jspdf = new jsPDF("p", "mm", "a4");
      const doc = vm.jspdf;

      const marginX = 0; //(pageWidth - canvasWidth) / 2;
      const marginY = 0; //(pageHeight - canvasHeight) / 2;

      const pageWidth = doc.internal.pageSize.getWidth();
      const pageHeight = doc.internal.pageSize.getHeight();
      const widthRatio = pageWidth / canvas.width;
      const heightRatio = pageHeight / canvas.height;
      const ratio = widthRatio > heightRatio ? heightRatio : widthRatio;
      const canvasWidth = canvas.width * ratio;
      const canvasHeight = canvas.height * ratio;

      const image = vm.getDataURL(canvas);
      this.jspdf.addImage(
        image,
        "JPEG",
        marginX,
        marginY,
        canvasWidth,
        canvasHeight,
        null,
        "SLOW"
      );

      const fileName = `Carlos_Soriano_${moment().format(
        "YYYY_MM_DD_HH_mm"
      )}.pdf`;
      this.jspdf.save(fileName);
    },
    async savePDF() {
      const vm = this;
      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true
      };
      vm.setCanvas(options, vm.toPDF);
    },
    // https://stackoverflow.com/questions/24069124/how-to-save-a-image-in-multiple-pages-of-pdf-using-jspdf
    // https://stackoverflow.com/questions/19272933/jspdf-multi-page-pdf-with-html-renderer/34934497#34934497
    // insterstingly this used to work even when pdf was not reassigned to paginated
    // eg
    // vm.paginate(target, canvas, pdf);
    // paginated.save(`Carlos_Soriano_${Date.now()}.pdf`);
    // vm.$emit('toPDF', false);
    async toPage() {
      const vm = this;
      const options = {
        width: 810,
        scale: 5,
        useCORS: true,
        allowTaint: true
      };
      vm.$emit("to-render-pdf", true);
      vm.setCanvas(options, vm.paginate);
    },
    paginate() {
      const canvas = this.canvas;
      this.jspdf = new jsPDF("p", "pt", "a4");
      const vm = this;
      const imgData = vm.getDataURL(canvas);

      const marginX = 0;
      const marginY = 0;

      const imgWidth = 595;
      const pageHeight = 842;
      let imgHeight = (canvas.height * imgWidth) / canvas.width;
      let heightLeft = imgHeight;
      this.jspdf.addImage(
        imgData,
        "JPEG",
        marginX,
        marginY,
        imgWidth,
        imgHeight,
        null,
        "SLOW"
      );

      heightLeft -= pageHeight;
      let secondHalf = heightLeft - imgHeight;

      // colorLog(`secondHalf: ${secondHalf}`, "violet");

      this.jspdf.addPage();
      // TODO
      // add multiple page logic
      this.jspdf.addImage(
        imgData,
        "PNG",
        0,
        secondHalf,
        imgWidth,
        imgHeight,
        null,
        "SLOW"
      );
      const fileName = `Carlos_Soriano_${moment().format(
        "YYYY_MM_DD_HH_mm"
      )}.pdf`;
      
      this.jspdf.save(fileName);

      vm.$emit("to-render-pdf", false);
      return this.jspdf;
    },
    /* istanbul ignore next */
    createDoc(canvas) {
      const doc = new jsPDF("p", "mm", "a4");
      const pageWidth = doc.internal.pageSize.getWidth();
      const pageHeight = doc.internal.pageSize.getHeight();
      const widthRatio = pageWidth / canvas.width;
      const heightRatio = pageHeight / canvas.height;
      const ratio = widthRatio > heightRatio ? heightRatio : widthRatio;
      const canvasWidth = canvas.width * ratio;
      const canvasHeight = canvas.height * ratio;
      const marginX = 0; //(pageWidth - canvasWidth) / 2;
      const marginY = 0; //(pageHeight - canvasHeight) / 2;
      return {
        doc,
        marginX,
        marginY,
        canvasWidth,
        canvasHeight
      };
    },
    /* istanbul ignore next */
    async _toPDF(canvas) {
      const vm = this;
      const image = vm.getDataURL(canvas);
      const { doc, marginX, marginY, canvasWidth, canvasHeight } = vm.createDoc(
        canvas
      );
      doc.addImage(
        image,
        "JPEG",
        marginX,
        marginY,
        canvasWidth,
        canvasHeight,
        null,
        "SLOW"
      );
      return doc;
    },
    /* istanbul ignore next */
    async _savePDF() {
      const vm = this;
      log("blue", "savePDF from Button Float");
      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true
      };
      try {
        const callback = async () => {
          const doc = await vm.toPDF(this.canvas);
          log("green", "before save");
          const fileName = `Carlos_Soriano_${Date.now()}.pdf`;
          doc.save(fileName);
          return {
            doc,
            fileName
          };
          // return Promise.resolve({
          //   doc,
          //   fileName
          // })
        };
        log("cyan", "about to do callback");
        vm.setCanvas(options, callback);
      } catch (err) {
        console.log(err);
      }
    }
  },
  mounted() {}
};
</script>

<style lang="scss">
</style>