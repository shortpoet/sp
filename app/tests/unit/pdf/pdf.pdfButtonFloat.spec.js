// TODO organize rest of tests' imports like this

// test utils
import * as myTest from "@/utils/myTest";
const { createWrapper } = myTest.default;
import { createLocalVue } from "@vue/test-utils";

// utils
import { cloneDeep } from "lodash";
import Vue from "vue";
import { log } from "@/utils/colorLog";

// plugins
import PortalVue from "portal-vue";
import {
  FontAwesomeIcon,
  FontAwesomeLayers,
} from "@fortawesome/vue-fontawesome";
import { library } from "@fortawesome/fontawesome-svg-core";
import {
  faCircle,
  faFilePdf,
  faSave,
  faTimes,
  faRocket,
} from "@fortawesome/free-solid-svg-icons";
import moment from "moment";

// SUT
import PDFButtonFloat from "@/components/Resume/PDF/PDFButtonFloat";
import PDF from "@/views/PDF";
import {hardResume} from "@/assets/resume.js";

// mocks
// the first way wasn't being 'called'
// it was loggin as a mock function but not the one being used in
// wrapper
// import jspdf from '@/__mocks__/jspdf'
import jspdf, { addImage } from "jspdf";
jest.mock("jspdf");
// similar to mocking @vue/testutils
// folders had to be nested just like source module for mock to work

// importing the mocks directly didn't seem to work
// but could have been due to other error
// consider checking
import html2canvas from "@trainiac/html2canvas";
// import html2canvas from '@/__mocks__/@trainiac/html2canvas'
import fontfaceobserver from "fontfaceobserver";
// import fontfaceobserver from '@/__mocks__/fontfaceobserver'

jest.useFakeTimers();

let wrapper;

describe("pdf.pdfButtonFloat", () => {
  const component = PDFButtonFloat;

  let methods = component.methods;

  const LocalVue = createLocalVue();
  LocalVue.use(PortalVue);

  let propsData;

  let computed;

  let stubs;

  let mocks;

  let ignoredElements;

  let resumeStoreOptions;

  let mountOptions;

  let expected;

  const events = [
    "click",
    "touchstart",
    "touchcancel",
    "touchmove",
    "touchend",
  ];

  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();

    jspdf.mockClear();

    jest.useFakeTimers();

    expected = ``;

    library.add(faCircle, faFilePdf, faSave, faTimes, faRocket);

    stubs = {
      FontAwesomeIcon,
      FontAwesomeLayers,
    };

    propsData = {
      target: "save-button-float",
      icon: "save",
      pdfTarget: "pdf-anchor",
    };

    computed = {};

    ignoredElements = [
      // 'portal'
    ];

    mocks = {};

    mountOptions = {
      propsData: propsData,
      stubs: stubs,
    };

    const getResumeLoaded = true;

    resumeStoreOptions = {
      getters: {
        getResumeLoaded: jest.fn(() => getResumeLoaded),
        getResume: jest.fn(() => hardResume),
      },
      mutations: { SET_RESUME_RAW: jest.fn() },
    };

  });

  describe("pdf.pdfButtonFloat.snapshot", () => {
    it("should match snapshot", () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe("pdf.pdfButtonFloat.showModal", () => {
    it("should show modal by changing isModalVisible to true", () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);
      expect(wrapper.vm.isModalVisible).toBe(false);
      // 1st error being thrown seemed to be from test thinking this.isModalVisible was a function assignment
      // fixed by changing to .call()
      // then error bec no this context in test to added wrapper.vm
      // https://medium.com/@lachlanmiller_52885/data-and-interaction-testing-in-vue-e7914a9179d7
      methods.showModal.call(wrapper.vm);
      expect(wrapper.vm.isModalVisible).toBe(true);
    });
  });

  describe("pdf.pdfButtonFloat.closeModal", () => {
    it("should close modal by changing isModalVisible to false", async () => {
      const data = () => {
        return {
          isModalVisible: true,
        };
      };
      mountOptions = {
        propsData: propsData,
        data: data,
      };

      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      expect(wrapper.vm.isModalVisible).toBe(true);
      methods.closeModal.call(wrapper.vm);
      expect(wrapper.vm.isModalVisible).toBe(false);
    });
  });

  describe("pdf.pdfButtonFloat.checkFonts", () => {
    it("should call fontfaceobserver", async () => {
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);
      await methods.checkFonts.call(wrapper.vm);
      expect(fontfaceobserver).toHaveBeenCalled();
    });
  });

  describe("pdf.pdfButtonFloat.getCanvas", () => {
    it("should call html2canvas with options", async () => {
      mocks = {
        dispatch: jest.fn(),
      };

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true,
      };

      // had to create wrapper from parent component because using vue portal
      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions);

      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true,
      };

      // this wasy doesn't load the context needed for method to run correctly
      // await methods.getCanvas(options)

      await wrapper.find(PDFButtonFloat).vm.getCanvas(options);

      const target = document.getElementById(
        wrapper.find(PDFButtonFloat).vm.pdfTarget
      );

      expect(html2canvas).toHaveBeenCalledWith(target, options);
    });
  });

  describe("pdf.pdfButtonFloat.createDoc", () => {
    it("should return doc with dimensions", async () => {
      mocks = {
        dispatch: jest.fn(),
      };

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true,
      };

      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions);

      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true,
      };

      // https://github.com/hustcc/jest-canvas-mock
      const getDataURL = jest.fn(
        () =>
          "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
      );

      wrapper.find(PDFButtonFloat).setMethods({ getDataURL: getDataURL });

      const canvas = await wrapper.find(PDFButtonFloat).vm.getCanvas(options);

      const createDoc = await wrapper.find(PDFButtonFloat).vm.createDoc(canvas);
      const doc = new jspdf("p", "mm", "a4");

      expect(createDoc).toMatchObject({
        doc: doc,
        marginX: 0,
        marginY: 0,
        canvasWidth: 888,
        canvasHeight: 888,
      });
    });
  });

  describe("pdf.pdfButtonFloat.setCanvas", () => {
    it("calls getCanvas assigning returned value to component data after 250s timeout", async () => {
      mocks = {
        dispatch: jest.fn(),
      };

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true,
      };

      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions);

      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true,
      };

      const getDataURL = jest.fn(
        () =>
          "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
      );
      wrapper.find(PDFButtonFloat).setMethods({ getDataURL: getDataURL });

      const callback = jest.fn();

      await wrapper.find(PDFButtonFloat).vm.setCanvas(options, callback);

      const canvas = { width: 888, height: 888 };

      expect(wrapper.find(PDFButtonFloat).vm.canvas).toBeNull();

      jest.advanceTimersByTime(250);
      await Promise.resolve();

      expect(wrapper.find(PDFButtonFloat).vm.canvas).toMatchObject(canvas);
    });
  });

  describe("pdf.pdfButtonFloat.toPDF", () => {
    const getDataURL = jest.fn(
      () =>
        "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
    );

    // need to mock canvas because this function expects it
    // to already be set by setcanvas which calls this as callback
    const canvas = { width: 888, height: 888 };

    const doc = new jspdf("p", "mm", "a4");
    
    const canvases = [ { width: 555, height: 888 }, { width: 888, height: 888 }, { width: 888, height: 555 } ]

    beforeEach(() => {
      // need before each else the stubs don't get reset (props among other things)

      jest.resetModules();
      jest.clearAllMocks();

      jspdf.mockClear();

      mocks = {
        dispatch: jest.fn(),
      };
      const data = () => {
        return {
          canvas: canvas,
        };
      };

      mountOptions = {
        propsData: propsData,
        data: data,
        stubs: stubs,
      };

      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      wrapper.setMethods({ getDataURL: getDataURL });
    });

    it("jspdf null on load and should create new jspdf and set it to data property", async () => {
      expect(wrapper.vm.jspdf).toBeNull();

      await wrapper.vm.toPDF();

      expect(wrapper.vm.jspdf).toMatchObject(doc);
    });

    canvases.forEach(canvas => {
      it(`should add image with correct dimensions w: ${canvas.width} & h: ${canvas.height}`, async () => {
        
        // must have fresh data object here so loaded SUT matches expected
        const data = () => {
          return {
            canvas: canvas,
          };
        };
  
        mountOptions = {
          propsData: propsData,
          data: data,
          stubs: stubs,
        };
  
        wrapper = createWrapper(component, mountOptions, resumeStoreOptions);
  
        wrapper.setMethods({ getDataURL: getDataURL });
          await wrapper.vm.toPDF();
  
        const addImageSpy = jest
          .spyOn(wrapper.vm.jspdf, "addImage")
          .mockImplementation(() => jest.fn());
  
        const marginX = 0; //(pageWidth - canvasWidth) / 2;
        const marginY = 0; //(pageHeight - canvasHeight) / 2;
  
        const pageWidth = doc.internal.pageSize.getWidth();
        const pageHeight = doc.internal.pageSize.getHeight();
        const widthRatio = pageWidth / canvas.width;
        const heightRatio = pageHeight / canvas.height;
        const ratio = widthRatio > heightRatio ? heightRatio : widthRatio;
        const canvasWidth = canvas.width * ratio;
        const canvasHeight = canvas.height * ratio;
  
        const image = getDataURL();
    
        expect(addImageSpy).toHaveBeenCalledWith(
          image,
          "JPEG",
          marginX,
          marginY,
          canvasWidth,
          canvasHeight,
          null,
          "SLOW"
        );
      });
    })


    it("should save pdf with correct filename", async () => {
      await wrapper.vm.toPDF();

      const saveSpy = jest
        .spyOn(wrapper.vm.jspdf, "save")
        .mockImplementation(() => jest.fn());

      const fileName = `Carlos_Soriano_${moment().format(
        "YYYY_MM_DD_HH_mm"
      )}.pdf`;

      expect(saveSpy).toHaveBeenCalledWith(fileName);
    });
  });

  describe("pdf.pdfButtonFloat.savePDF", () => {
    it("should call setCanvas with options and pdf callback", async () => {
      mocks = {
        dispatch: jest.fn(),
      };

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true,
      };

      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions);

      const options = {
        scale: 5,
        useCORS: true,
        allowTaint: true,
      };

      const getDataURL = jest.fn(
        () =>
          "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
      );

      const setCanvas = jest.fn();

      const toPDF = jest.fn();

      // i really like this test, it doesn't let me arbitrarily change method names! XD
      wrapper.find(PDFButtonFloat).setMethods({
        getDataURL: getDataURL,
        setCanvas: setCanvas,
        toPDF: toPDF,
      });

      await wrapper.find(PDFButtonFloat).vm.savePDF();

      wrapper.find(PDFButtonFloat).setMethods({ getDataURL: getDataURL });
      expect(setCanvas).toHaveBeenCalledWith(options, toPDF);
    });
  });
  describe("pdf.pdfButtonFloat.paginate", () => {

    const getDataURL = jest.fn(
      () =>
        "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
    );

    const doc = new jspdf("p", "mm", "a4");

    beforeEach(() => {
      // need before each else the stubs don't get reset (props among other things)
  
      // need to mock canvas because this function expects it
      // to already be set by setcanvas which calls this as callback
      const canvas = { width: 888, height: 888 };
    
      const data = () => {
        return {
          canvas: canvas,
        };
      };

      mountOptions = {
        propsData: propsData,
        data: data,
        stubs: stubs,
      };


      jest.resetModules();
      jest.clearAllMocks();

      jspdf.mockClear();

      mocks = {
        dispatch: jest.fn(),
      };

      // a typo in one of the paginate methods which called toPDF instead
      // created an error only when running the whole file instead of just the describe

      });

    it("jspdf null on load and should create new jspdf and set it to data property", async () => {
      // added new addPage method to mock
      
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      wrapper.setMethods({ getDataURL: getDataURL });
  
      expect(wrapper.vm.jspdf).toBeNull();

      await wrapper.vm.paginate();

      log('green', wrapper.vm.jspdf)

      expect(wrapper.vm.jspdf).toMatchObject(doc);
    });

    it("should add image with correct dimensions and divides up the page into chunks that fit an a4 then switches back to regular layout", async () => {
      
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      wrapper.setMethods({ getDataURL: getDataURL });
  
      const canvas = { width: 888, height: 888 };
    
      await wrapper.vm.paginate();

      log('green', wrapper.vm.jspdf)

      const addImageSpy = jest
        .spyOn(wrapper.vm.jspdf, "addImage")
        .mockImplementation(() => jest.fn());

      const marginX = 0; //(pageWidth - canvasWidth) / 2;
      const marginY = 0; //(pageHeight - canvasHeight) / 2;

      const imgWidth = 595;
      const pageHeight = 842;
      let imgHeight = (canvas.height * imgWidth) / canvas.width;
      let heightLeft = imgHeight;

      const imgData = getDataURL();

      expect(addImageSpy).toHaveBeenCalledWith(
        imgData,
        "JPEG",
        marginX,
        marginY,
        imgWidth,
        imgHeight,
        null,
        "SLOW"
      );

      log("green", "test");

      expect(wrapper.find(PDFButtonFloat).emitted()).toMatchObject({
        "to-render-pdf": [[false]],
      });
    });

    it("should save pdf with correct filename", async () => {
      
      wrapper = createWrapper(component, mountOptions, resumeStoreOptions);

      wrapper.setMethods({ getDataURL: getDataURL });
        
      await wrapper.vm.paginate();

      log('green', wrapper.vm.jspdf)

      const saveSpy = jest
        .spyOn(wrapper.vm.jspdf, "save")
        .mockImplementation(() => jest.fn());

      const fileName = `Carlos_Soriano_${moment().format(
        "YYYY_MM_DD_HH_mm"
      )}.pdf`;

      expect(saveSpy).toHaveBeenCalledWith(fileName);
    });

  });

  describe("pdf.pdfButtonFloat.toPage", () => {
    it("should call setCanvas with options and paginate callback after emitting to-render-pdf", async () => {
      mocks = {
        dispatch: jest.fn(),
      };

      mountOptions = {
        propsData: propsData,
        mocks: mocks,
        stubs: stubs,
        attachToDocument: true,
      };

      wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions);

      const options = {
        width: 810,
        scale: 5,
        useCORS: true,
        allowTaint: true,
      };

      const getDataURL = jest.fn(
        () =>
          "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=="
      );

      const setCanvas = jest.fn();

      const paginate = jest.fn();

      wrapper.find(PDFButtonFloat).setMethods({
        getDataURL: getDataURL,
        setCanvas: setCanvas,
        paginate: paginate,
      });

      await wrapper.find(PDFButtonFloat).vm.toPage();

      expect(wrapper.find(PDFButtonFloat).emitted()).toMatchObject({
        "to-render-pdf": [[true]],
      });

      expect(setCanvas).toHaveBeenCalledWith(options, paginate);
    });
  });

  describe('#### _toPDF ####', () => {
    it("#### was a crazy test for some spaghetti functions ####", async () => {
      //   mocks = {
      //     dispatch: jest.fn()
      //   }
      //   mountOptions = {
      //     propsData: propsData,
      //     mocks: mocks,
      //     stubs: stubs,
      //     attachToDocument: true
      //   }
      //   // had to create wrapper from parent component because using vue portal
      //   wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions)
      //   const options = {
      //     scale: 5,
      //     useCORS: true,
      //     allowTaint: true,
      //   }
      //   // this wasy doesn't load the context needed for method to run correctly
      //   // await methods.getCanvas(options)
      //   // const getCanvas = jest.fn(() => 'data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==')
      //   const getDataURL = jest.fn(() => 'data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==')
      //   wrapper.find(PDFButtonFloat).setMethods({getDataURL: getDataURL})
      //   // let savePDF = await wrapper.find(PDFButtonFloat).vm.savePDF()
      //   // console.log(savePDF)
      //   // const doc = new jspdf('p', 'mm', 'a4');
      //   // expect(savePDF).toBeUndefined()
      //   // expect(jspdf).toHaveBeenCalledTimes(1)
      //   // // jest.advanceTimersByTime(250)
      //   // // await Promise.resolve()
      //   // console.log(savePDF)
      //   // const fileName = `Carlos_Soriano_${Date.now()}.pdf`
      //   // expect(savePDF).toMatchObject({doc, fileName})
      //   // jest.advanceTimersByTime(250)
      //   return wrapper.find(PDFButtonFloat).vm.savePDF().then( async (first) => {
      //     jest.advanceTimersByTime(250)
      //     await Promise.resolve()
      //     console.log(first)
      //   }).then(async (next) => {
      //     jest.advanceTimersByTime(250)
      //     await Promise.resolve()
      //     // this gets called right after timeout
      //     // and before callback
      //     console.log(next)
      //   }).then(async (next2) => {
      //     jest.advanceTimersByTime(250)
      //     await Promise.resolve()
      //     // by now doc has been logged from callback
      //     console.log(next2)
      //   }).then(async (next3) => {
      //     jest.advanceTimersByTime(250)
      //     await Promise.resolve()
      //     console.log(next3)
      //   })
    });

    it('#### should call jsPDF with options ####', async () => {

      // mocks = {
      //   dispatch: jest.fn()
      // }

      // mountOptions = {
      //   propsData: propsData,
      //   mocks: mocks,
      //   stubs: stubs,
      //   attachToDocument: true
      // }

      // wrapper = createWrapper(PDF, mountOptions, resumeStoreOptions)

      // const options = {
      //   scale: 5,
      //   useCORS: true,
      //   allowTaint: true,
      // }

      // const getDataURL = jest.fn(() => 'data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==')
      // wrapper.find(PDFButtonFloat).setMethods({getDataURL: getDataURL})

      // const doc = new jspdf('p', 'mm', 'a4');

      // expect(jspdf).toHaveBeenCalledTimes(1)

      // const canvas = await wrapper.find(PDFButtonFloat).vm.getCanvas(options)

      // const toPDF = await wrapper.find(PDFButtonFloat).vm.toPDF(canvas)

      // expect(toPDF).toMatchObject(doc)

    })
  })
});
