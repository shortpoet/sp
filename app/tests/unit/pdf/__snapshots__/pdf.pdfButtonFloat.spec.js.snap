// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pdf.pdfButtonFloat pdf.pdfButtonFloat.snapshot should match snapshot 1`] = `<div class="v-portal" style="display: none;"></div>`;
